sys/include/net/nanocoap.h

```diff
@@ -1099,6 +1099,15 @@ static inline unsigned coap_size2szx(unsigned len)
  * full.
  */
 /**@{*/
+
+/**
+ * @brief   Get next consecutive message ID for use when building a new
+ *          CoAP request.
+ *
+ * @return  A new message ID that can be used for a request or response.
+ */
+uint16_t coap_next_msg_id(void);
```

Comment:
Could you change so that the src and dst endpoint are available to the implementation?

The standard suggest three levels of granularity:

1. Having one counter for each distinct src <--> dst pair
2. Having one counter per network interface / subnet /
3. Having one single globel counter

The reason for having more counter is that message IDs must not be reused within some time. It easily becomes the limiting factor on 802.3 or 802.11 networks. So, having multiple counters can increase throughput.

I think it is fine to keep using a single counter. But having a way to extend would be nice.

Reply:
Then we better keep the separate implementations.
With `nanocoap_sock` it's trivial to attach such a counter to the `nanocoap_sock_t` struct.

If we want a generalized function we would need to work with arrays to look up the message ID, which sounds like a disproportionate overhead to me.